packages/suite/src/actions/suite/metadataLabelingActions.ts

```diff
@@ -529,6 +568,297 @@ export const addMetadata =
+            await dispatch(getMetadataFiles());
+
+        // 4. there are no old files (either labeling was never used before, old old files were renamed to file_v1.mdtd)
+        // also note, that we take into account only those oldEncryption files which do not have their renamed version concurrently existinging in renamedOldEncryptionFiles
+        // this could happen in a very rare edgecase:
+        // 1. user does migration in updated suite which already has encryption v2, this renames old encryption file
+        // 2. user goes to old suite, creates a label, old encryption file is created. now old renamed and old exist together
+        // 3. => this means that suite is trying to "enable labeling" forever but it never actually carries out migration because it stops on later
+        //       condition "everyEntityHasNewFile"
+        if (
+            oldEncryptionFiles.filter(
+                file => !renamedOldEncryptionFiles.includes(`${file}_v1.mtdt`),
+            ).length === 0
+        ) {
+            return { success: true };
+        }
+
+        // 5. there are old files, but also all labelable entities currently known to suite have some record in currentEncryptionFiles.
+        //    this means that they have already been migrated or dummy file was created
+        const everyEntityHasNewFile = dispatch(getLabelableEntities(deviceState)).every(entity => {
+            const nextKeys = entity[2];
+
+            if (!nextKeys) {
+                // todo: should never happend but happend once during testing.
+                throw new Error('metadata migration: next keys are missing');
+            }
+
+            return currentEncryptionFiles.find(file => file === nextKeys.fileName);
+        });
+
+        if (everyEntityHasNewFile) {
+            return { success: true };
+        }
+
+        // 7. sync metadata keys for prev encryption version.
+        //    NOTE: result of this operation is saved for device and account encryption keys are computed from it. This means that we can add new accounts at any point of time later without calling this again
+        if (!device.metadata[prevEncryptionVersion]) {
+            await dispatch(setDeviceMetadataKey(device, prevEncryptionVersion));
+        }
+        device = selectDeviceByState(getState(), deviceState);
+        if (!device?.metadata[prevEncryptionVersion]?.key) {
+            return { success: false, error: 'metadata migration: cancelled' };
+        }
+
+        dispatch(syncMetadataKeys(device, prevEncryptionVersion));
+
+        if (!device.state) {
+            // this should never happen
+            return { success: false, error: 'metadata migration: device not authorized' };
+        }
+
+        // 8. get labelable entitites again (there was async operation in between)
+        const allEntities = dispatch(getLabelableEntities(device.state));
+
+        // 9. split labelable entities into 2 groups:
```

Comment:
This, but also many other steps shall be separated into own well named functions (and probably unit-tested).

This would make code less spaghetti-like, easier to read and named well named function will reduce the need to long comments significantly